File: A2209I/MVC/de07/Controllers/UserController.cs
using de07.Data;
using de07.Models;
using de07.ViewModels;
using Microsoft.AspNetCore.Identity;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace de07.Controllers
{
    public class UserController : Controller
    {        
        private readonly UserManager<ApplicationUser> _userManager;
        private readonly SignInManager<ApplicationUser> _signInManager;
        private readonly ApplicationDbContext _applicationDbContext;
        

        public UserController(
            UserManager<ApplicationUser> userManager, 
            SignInManager<ApplicationUser> signInManager,
            ApplicationDbContext applicationDbContext
            )
        {
            _userManager = userManager;
            _signInManager = signInManager;
            _applicationDbContext = applicationDbContext;
        }

        [HttpGet]        
        public async Task<IActionResult> Login()
        {
            return View(new LoginViewModel()); // Initializes a new instance for the view
        }

        [HttpPost]
        [ValidateAntiForgeryToken]
        public async Task<IActionResult> Login(LoginViewModel model)
        {            
            if (ModelState.IsValid)
            {
                var user = await _applicationDbContext.Users
                    .Where(item => item.Email.Equals(model.Email))
                    .FirstOrDefaultAsync();
                if (user != null)
                {
                    // Attempt to sign in the user
                    var result = await _signInManager.PasswordSignInAsync(user, model.Password, model.RememberMe, lockoutOnFailure: false);

                    if (result.Succeeded)
                    {
                        return RedirectToAction(nameof(ServiceController.Index), "Service");
                    }

                    if (result.IsLockedOut)
                    {
                        // If the user is locked out, handle accordingly
                        ModelState.AddModelError(string.Empty, "This account has been locked out. Please try again later.");
                    }
                    else
                    {
                        // Add error message if login attempt fails
                        ModelState.AddModelError(string.Empty, "Invalid login attempt.");
                    }
                }
                else
                {
                    ModelState.AddModelError(string.Empty, "No account found with this email address.");
                }
                              
            }

            // Nếu không thành công, hiển thị lại form
            return View(model);
        }
        [HttpGet]
        public IActionResult Register()
        {
            return View(new RegisterViewModel());  // This will return the Register view
        }
        [HttpPost]
        [ValidateAntiForgeryToken]
        public async Task<IActionResult> Register(RegisterViewModel model)
        {
            if (ModelState.IsValid)
            {
                var user = new ApplicationUser
                {
                    UserName = model.Email,
                    Email = model.Email,
                    AddressHome = model.AddressHome,
                    PhoneHome = model.PhoneHome,
                    Job = model.Job
                };
                var result = await _userManager.CreateAsync(user, model.Password);
                if (result.Succeeded)
                {
                    // Optionally sign-in the user after registration
                    await _signInManager.SignInAsync(user, isPersistent: false);
                    
                    // Redirect to login or home page after successful registration
                    return RedirectToAction("Index", "Service");
                }
                foreach (var error in result.Errors)
                {
                    ModelState.AddModelError(string.Empty, error.Description);
                }
            }

            // Something failed, redisplay form
            return View(model);
        }


    }
}

File: C2110iDoiCan/IASF/de01/src/main/java/com/example/de01/controllers/UserController.java
package com.example.de01.controllers;

import com.example.de01.services.user.UserService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import com.example.de01.models.*;
import org.springframework.ui.Model;
@Controller
@RequestMapping("/users")
@RequiredArgsConstructor
public class UserController {
    private final UserService userService;

    @GetMapping("/login")
    public String login() {
        return "user/login";//where is path: resources/???
    }
    //sau khi bam login=
    @PostMapping("/login")
    public String handleLogin(
            @RequestParam("username") String username,
            @RequestParam("password") String password,
            Model model) {
        User user = userService.login(username, password);
        if (user != null) {
            return "redirect:/home";  // Redirect to the home page after successful login
        } else {
            model.addAttribute("errorMessage",
                    "Please enter correct username and password to login");
            return "user/login";  // Return to the login page with an error message
        }
    }
}
